src/js/integrations/stalltracking.ts

```diff
@@ -0,0 +1,328 @@
+
+    return (endTimestamp?: number) =>
+      this._onFinish(transaction, statsOnStart, endTimestamp);
+  }
+
+  /**
+   * Logs a transaction as finished.
+   * Stops stall tracking if no more transactions are running.
+   * @returns The stall measurements
+   */
+  private _onFinish(
+    transaction: Transaction,
+    statsOnStart: StallMeasurements,
+    passedEndTimestamp?: number
+  ): StallMeasurements | null {
+    const _endTimestamp = passedEndTimestamp ?? transaction.endTimestamp;
+    const endTimestamp = _endTimestamp
+      ? standardizeTimestampToSeconds(_endTimestamp)
+      : undefined;
+
+    const statsOnFinish = endTimestamp
+      ? this._findNearestSpanEnd(transaction.spanId, endTimestamp)
+      : this._getCurrentStats(transaction.spanId);
+
+    this._runningTransactions = this._runningTransactions.filter(
+      (t) => t.spanId !== transaction.spanId
+    );
+    this._longestStallsByTransaction.delete(transaction.spanId);
+    this._statsAtTimestamp.delete(transaction.spanId);
+
+    if (this._runningTransactions.length === 0) {
+      this._stopTracking();
+    }
+
+    if (!statsOnFinish) {
+      if (typeof _endTimestamp !== "undefined") {
+        logger.log(
+          "[StallTracking] Stall measurements not added due to `endTimestamp` being set to a value too far away from a logged point."
+        );
+      } else {
+        logger.log(
+          "[StallTracking] Stall measurements not added due to not being able to determine the end timestamp."
+        );
+      }
+
+      return null;
+    }
+
+    return {
+      stall_count: {
+        value: statsOnFinish.stall_count.value - statsOnStart.stall_count.value,
+      },
+      stall_total_time: {
+        value:
+          statsOnFinish.stall_total_time.value -
+          statsOnStart.stall_total_time.value,
+      },
+      stall_longest_time: statsOnFinish.stall_longest_time,
+    };
+  }
+
+  /**
+   * `spanEndTimestamp` needs to be in seconds!
+   */
+  private _logSpanFinish(
+    transactionId: string,
+    spanEndTimestamp: number
+  ): void {
+    const statsAtTimestampPerTransaction = this._statsAtTimestamp.get(
+      transactionId
+    );
+
+    if (statsAtTimestampPerTransaction) {
+      if (
+        Math.abs(timestampInSeconds() - spanEndTimestamp) >
+        MARGIN_OF_ERROR_SECONDS
+      ) {
+        logger.log(
+          "[StallTracking] Span end not logged due to end timestamp being outside the margin of error from now."
+        );
+      } else {
+        statsAtTimestampPerTransaction.set(
+          spanEndTimestamp,
+          this._getCurrentStats(transactionId)
+        );
+      }
+
+      // We delete the first span (earliest timestamp) if there are greater than N span stats in the map.
+      if (statsAtTimestampPerTransaction.size > MAX_SPAN_STATS_LEN) {
+        const [key] = statsAtTimestampPerTransaction.keys();
+        statsAtTimestampPerTransaction.delete(key);
```

Comment:
Nice, forgot that maps keep insertion order

Reply:
Haha yeah when I was about to write this and go through the keys I was like wait, they're already in insertion order lol